app/Dockerfile

```diff
@@ -4,34 +4,123 @@
+    postgresql \
+    # Reduce the image size by clear apt cached lists
+    # Complies with https://github.com/codacy/codacy-hadolint/blob/master/codacy-hadolint/docs/description/DL3009.md
+    && rm -fr /var/lib/apt/lists/* \
+    && rm /etc/ssl/private/ssl-cert-snakeoil.key
 
-# Set the application working directory.
-WORKDIR /srv
+ARG RUN_UID
+ARG RUN_USER
+
+# The following logic creates the RUN_USER home directory and the directory where
+# we will be storing the application in the image. This runs when the user is not root
+RUN : "${RUN_USER:?RUN_USER and RUN_UID need to be set and non-empty.}" && \
+    [ "${RUN_USER}" = "root" ] || \
+    (useradd --create-home --create --user-group --home "/home/${RUN_USER}" --uid ${RUN_UID} "${RUN_USER}" \
+    && mkdir /app \
+    && chown -R ${RUN_UID} "/home/${RUN_USER}" /app)
+
+#-----------
+# Dev image
+#-----------
+
+FROM base AS dev
+ARG RUN_USER
+USER ${RUN_USER}
+WORKDIR /app
 
 COPY pyproject.toml poetry.lock ./
+# Explicitly create a new virtualenv to avoid getting overridden by mounted .venv folders
 RUN poetry config virtualenvs.in-project false && poetry env use python
-RUN poetry install --no-root
+# Install all dependencies including dev dependencies
+RUN poetry install --no-root --with dev
 
-# Copy application files.
-COPY . /srv
+COPY . /app
 
 # Set the host to 0.0.0.0 to make the server available external
 # to the Docker container that it's running in.
 ENV HOST=0.0.0.0
 
-# Install application dependencies.
-# https://python-poetry.org/docs/basic-usage/#installing-dependencies
-RUN poetry install
 # Run the application.
 CMD ["poetry", "run", "python", "-m", "src"]
+
+#---------------
+# Release Build
+#---------------
+
+FROM base AS release-build
+
+WORKDIR /app
+
+COPY . /app
+
+# Remove any existing virtual environments that might exist. This
+# might happen if testing out building the release image from a local machine
+# that has a virtual environment within the project app folder.
+RUN rm -fr /app/.venv
+
+# Set virtualenv location to be in project to be easy to find
+# This will create a virtualenv in /app/.venv/
+# See https://python-poetry.org/docs/configuration/#virtualenvsin-project
+# See https://python-poetry.org/docs/configuration/#using-environment-variables
+ENV POETRY_VIRTUALENVS_IN_PROJECT=true
+
+# Install production runtime dependencies only
+RUN poetry install --no-root --only main
+
+# Build the application binary (python wheel) defined in pyproject.toml
+# Note that this will only copy over python files, and files stated in the
+# include section in pyproject.toml. Also note that if you change the name or
+# version section in pyproject.toml, you will need to change the dist/... to match
+# or the application will not build
+RUN poetry build --format wheel && poetry run pip install 'dist/template_application_flask-0.1.0-py3-none-any.whl'
+
+
+#---------------
+# Release image
+#---------------
+
+FROM scratch AS release
+
+ARG RUN_USER
+USER ${RUN_USER}
```

Comment:
Small nit, but maybe throw a comment here saying that to run any commands that need to be as `root` should be ran here

Reply:
I removed the productionization of the Dockerfile from this PR since something wasn't working with the wget healthcheck and I realize that that wasn't strictly necessary anyways for what we're doing here, so to keep this PR smaller, we can leave the minimizing of the Docker image for this ticket: #23

Reply:
re: adding the comment here, i'm trying to avoid being too wordy with comments. i don't really think it's necessary to have many placeholder comments like "put X here"